File: kt/src/commonMain/kotlin/com/github/ephemient/aoc2022/Day18.kt
package com.github.ephemient.aoc2022

@Day
class Day18(lines: List<String>) {
    private val points = buildSet {
        lines.mapTo(this) { line ->
            val (x, y, z) = line.split(',', limit = 3)
            IntTriple(x.toInt(), y.toInt(), z.toInt())
        }
    }

    @Day.Part
    fun part1(): Int = points.sumOf { point -> point.neighbors().count { it !in points } }

    @Day.Part
    fun part2(): Int {
        var minX = Int.MAX_VALUE
        var maxX = Int.MIN_VALUE
        var minY = Int.MAX_VALUE
        var maxY = Int.MIN_VALUE
        var minZ = Int.MAX_VALUE
        var maxZ = Int.MIN_VALUE
        for ((x, y, z) in points) {
            if (x < minX) minX = x
            if (x > maxX) maxX = x
            if (y < minY) minY = y
            if (y > maxY) maxY = y
            if (z < minZ) minZ = z
            if (z > maxZ) maxZ = z
        }
        val outside = buildSet {
            val queue = mutableListOf(IntTriple(minX - 1, minY - 1, minZ - 1).also { add(it) })
            while (queue.isNotEmpty()) {
                for (neighbor in queue.removeLast().neighbors()) {
                    neighbor.first in minX - 1..maxX + 1 &&
                        neighbor.second in minY - 1..maxY + 1 &&
                        neighbor.third in minZ - 1..maxZ + 1 &&
                        neighbor !in points &&
                        add(neighbor) &&
                        queue.add(neighbor)
                }
            }
        }
        return points.sumOf { point -> point.neighbors().count { it in outside } }
    }
}

private fun IntTriple.neighbors() = listOf(
    copy(first = first - 1),
    copy(first = first + 1),
    copy(second = second - 1),
    copy(second = second + 1),
    copy(third = third - 1),
    copy(third = third + 1),
)

File: py/aoc2022/day18.py
"""
Day 18: Boiling Boulders
"""

SAMPLE_INPUT = [
    "2,2,2",
    "1,2,2",
    "3,2,2",
    "2,1,2",
    "2,3,2",
    "2,2,1",
    "2,2,3",
    "2,2,4",
    "2,2,6",
    "1,2,5",
    "3,2,5",
    "2,1,5",
    "2,3,5",
]


def _parse(lines):
    for line in lines:
        x, y, z = line.rstrip().split(",", 2)
        yield int(x), int(y), int(z)


def _neighbors(point):
    x, y, z = point
    return [
        (x - 1, y, z),
        (x + 1, y, z),
        (x, y - 1, z),
        (x, y + 1, z),
        (x, y, z - 1),
        (x, y, z + 1),
    ]


def part1(lines):
    """
    >>> part1(SAMPLE_INPUT)
    64
    """
    points = set(_parse(lines))
    return sum(
        neighbor not in points for point in points for neighbor in _neighbors(point)
    )


def part2(lines):
    """
    >>> part2(SAMPLE_INPUT)
    58
    """
    points = set(_parse(lines))
    min_x, max_x = min(x for x, _, _ in points), max(x for x, _, _ in points)
    min_y, max_y = min(y for _, y, _ in points), max(y for _, y, _ in points)
    min_z, max_z = min(z for _, _, z in points), max(z for _, _, z in points)
    outside = {(min_x - 1, min_y - 1, min_z - 1)}
    queue = list(outside)
    while queue:
        for point in _neighbors(queue.pop()):
            x, y, z = point
            if (
                min_x - 1 <= x <= max_x + 1
                and min_y - 1 <= y <= max_y + 1
                and min_z - 1 <= z <= max_z + 1
                and point not in points
                and point not in outside
            ):
                outside.add(point)
                queue.append(point)
    return sum(
        neighbor in outside for point in points for neighbor in _neighbors(point)
    )


parts = (part1, part2)
